client/landing/stepper/utils/performance-tracking.tsx

```diff
@@ -0,0 +1,53 @@
+import { useLayoutEffect, useMemo } from 'react';
+import { useStore } from 'react-redux';
+import {
+	startPerformanceTracking,
+	stopPerformanceTracking,
+} from 'calypso/lib/performance-tracking/lib';
+
+// Manually start performance tracking.
+// Used in the entry point to start data collection as close as possible to boot time.
+export function startStepperPerformanceTracking( {
+	fullPageLoad = false,
+}: {
+	fullPageLoad?: boolean;
+} ) {
+	startPerformanceTracking( 'stepper', { fullPageLoad } );
+}
+
+// This hook starts performance gathering for the Stepper everytime the flow or step change.
+// The library we're using ignores follow-up `start` calls while there's an inflight tracking
+// report, so this hook is always safe, even though the entry point starts its own report on
+// first render.
+export function useStartStepperPerformanceTracking( flow: string, step: string ) {
+	useMemo( () => {
+		startStepperPerformanceTracking( { fullPageLoad: false } );
+		// We need to start tracking again everytime we change flow or step.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [ flow, step ] );
+}
+
+function useStepperPerformanceTrackerStop( flow: string, step: string ) {
+	const performanceMetadata = useMemo( () => ( { flow: flow, step: step } ), [ flow, step ] );
+	const store = useStore();
+
+	// Use `useLayoutEffect` + rAF to be as close as possible to the actual rendering.
+	useLayoutEffect( () => {
+		requestAnimationFrame( () => {
+			stopPerformanceTracking( 'stepper', {
+				/// @ts-expect-error State is not properly typed in the performance tracking lib.
+				state: store.getState(),
+				metadata: performanceMetadata,
+			} );
+		} );
+	}, [ store, performanceMetadata ] );
+}
+
+// Stop performance data collection and report.
+// This component is placed at the end of the render tree to indicate completion.
+export function StepperPerformanceTrackerStop( { flow, step }: { flow: string; step: string } ) {
+	useStepperPerformanceTrackerStop( flow, step );
```

Comment:
I think making a component that doesn't render anything is a bit noisy. I would just call this hook where the component is used.

Reply:
It is important because we can't call hooks conditionally and we need it [here](https://github.com/Automattic/wp-calypso/pull/92130/files#diff-112b5e3b26ab85ab00ed2109fd109aeb8ea3f3a52916d847e4307c0225504d3fR58) 


Where we are only rendering the stop after some checkings.

Reply:
Yes, I figured, I would add `enabled` flag to `useStepperPerformanceTrackerStop` and add `if` inside the `useEffect`. Not a blocker by any means though.

Reply:
This is how the component is currently being used:

```js
...
  { stepContent }
  { stepContent && <StepperPerformanceTrackerStop flow={ flow.name } step={ step.slug } /> }
...
```

This ensures that the hook is only triggered when we have step content, and after it's done rendering, thus easily reusing all of the existing logic.

We could do this as a plain hook on `StepRoute`, but that would require reworking some of the existing logic, such as:

```js
...
  if ( shouldSkipRender ) {
    return null;
  }
...
```

given that hooks can't be inside conditionals. Ultimately I think the component is simpler, but let me know if you'd like me to make the change!

Reply:
Let's keep a component 👍🏼